Implement conditional expression changes in definite assignment

RikkiGibson opened the pull request:
Note that ed38577 simply deletes and regenerates the baseline in order to make comparison easier in the subsequent commit.

I decided to do this bit first, as the precise method of doing it had been in my head. There are corresponding nullable analysis changes I'd like to make, but nullable does more complex things with conditional expressions. Therefore I'd like to do those changes in a subsequent PR.

/cc @jcouv

Implementation of part of dotnet/csharplang#4472. The relevant spec sections implemented in this PR are:

> ## Boolean constant expressions
> We introduce a new section "Boolean constant expressions":
> 
> For an expression *expr* where *expr* is a constant expression with a bool value:
> - The definite assignment state of *v* after *expr* is determined by:
>   - If *expr* is a constant expression with value *true*, and the state of *v* before *expr* is "not definitely assigned", then the state of *v* after *expr* is "definitely assigned when false".
>   - If *expr* is a constant expression with value *false*, and the state of *v* before *expr* is "not definitely assigned", then the state of *v* after *expr* is "definitely assigned when true".
> 
> ### Remarks
> 
> We assume that if an expression has a constant value bool `false`, for example, it's impossible to reach any branch that requires the expression to return `true`. Therefore variables are assumed to be definitely assigned in such branches. This ends up combining nicely with the spec changes for expressions like `??` and `?:` and enabling a lot of useful scenarios.
> 
> It's also worth noting that we never expect to be in a conditional state *before* visiting a constant expression. That's why we do not account for scenarios such as "*expr* is a constant expression with value *true*, and the state of *v* before *expr* is "definitely assigned when true".
> 
> 
> ## ?: (conditional) expressions
> We augment the section [**?: (conditional) expressions**](../spec/variables.md#-conditional-expressions) as follows:
> 
> For an expression *expr* of the form `expr_cond ? expr_true : expr_false`:
> - ...
> - The definite assignment state of *v* after *expr* is determined by:
>   - ...
>   - If the state of *v* after *expr_true* is "definitely assigned when true", and the state of *v* after *expr_false* is "definitely assigned when true", then the state of *v* after *expr* is "definitely assigned when true".
>   - If the state of *v* after *expr_true* is "definitely assigned when false", and the state of *v* after *expr_false* is "definitely assigned when false", then the state of *v* after *expr* is "definitely assigned when false".
> 
> ### Remarks
> 
> This makes it so when both arms of a conditional expression result in a conditional state, we join the corresponding conditional states and propagate it out instead of unsplitting the state and allowing the final state to be non-conditional. This enables scenarios like the following:
> 
> ```cs
> bool b = true;
> object x = null;
> int y;
> if (b ? x != null && Set(out y) : x != null && Set(out y))
> {
>   y.ToString();
> }
> 
> bool Set(out int x) { x = 0; return true; }
> ```
> 
> This is an admittedly niche scenario, that compiles without error in the native compiler, but was broken in Roslyn in order to match the specification at the time. See [internal issue](http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/529603).

Relates to test plan https://github.com/dotnet/roslyn/issues/51463

RikkiGibson commented:
@dotnet/roslyn-compiler please review.

jcouv commented on src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:
```diff
@@ -2096,11 +2031,74 @@ static bool Set(out int x)
+    const bool FIELD_TRUE = true;
+    const bool FIELD_FALSE = false;
+
+    static void M()
```
I'd feel more comfortable using simple blocks to scope variables, rather than local functions (which introduce their own analysis quirks).

RikkiGibson replied:
Good idea. Done.

jcouv reviewed (approved):
LGTM Thanks (iteration 5)

333fred commented on src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:
```diff
@@ -1824,252 +1824,187 @@ public static void M8()
             CreateCompilation(source).VerifyDiagnostics(
-                // (61,58): error CS0165: Use of unassigned local variable 'a'
+                // (87,58): error CS0165: Use of unassigned local variable 'a'
                 // /* NDA --> NDA */       { int a; if (x               ? F(a) : F(b)) b = c; else d = c; } // Error
```
Are these inline comments all correct in the original code now?

RikkiGibson replied:
I believe so, but I'll make another pass to make sure. I actually suspect that they were originally written for a test in the native compiler and never updated to reflect the breaking change in Roslyn.

RikkiGibson replied:
By restoring native compiler behavior, a lot of the places that say OK in this test went from inaccurate to accurate again.￼

333fred replied:
Ok, so all the comments in here were pretty much just bad... got it.

RikkiGibson replied:
Just in case you were curious. There are 60 `// Error` comments in the test source, and now 60 calls to `Diagnostic` in the arguments to VerifyDiagnostics, and 60 `// Error` comments in the arguments to VerifyDiagnostics. So it looks like the comments are now accurate, FWIW.

333fred commented:
> /* DAT?NDA:NDA-->NDA */ { int a; if ((x && G(out a)) ? y               : z)               b = a; else d = c; } // Error

I think it would be worth adding some variations of these tests (where the condition is DAT/F) and the branch is a constant value. ie, something like `{ int a; if ((x && G(out a)) ? true               : z)               b = a; else d = c; } // OK`

---
Refers to: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:1752 in f126ebe. [](commit_id = f126ebee6bf841286dcd9ebd4ac8af8ae91c5bfe, deletion_comment = False)

333fred commented:
> /* DAT?NDA:NDA-->NDA */ { int a; if ((x && G(out a)) ? y               : z)               b = a; else d = c; } // Error

Also, is `!` support in this PR, or will it be in a future PR? We'll want to have variations that involve that.

---
In reply to: [789277456](https://github.com/dotnet/roslyn/pull/51498#issuecomment-789277456) [](ancestors = 789277456)

---
Refers to: src/Compilers/CSharp/Test/Semantic/FlowAnalysis/FlowTests.cs:1752 in f126ebe. [](commit_id = f126ebee6bf841286dcd9ebd4ac8af8ae91c5bfe, deletion_comment = False)

333fred commented:
Done review pass (commit 5)

RikkiGibson commented:
> Also, is ! support already implemented, or will it be in a future PR? We'll want to have variations that involve that.

Does this refer to unary not or to nullable suppression?

333fred commented:
> Does this refer to unary not or to nullable suppression?

Unary not.

RikkiGibson commented:
I have added a related test case in `IfConditionalConstant`. No implementation change needed. The existing handling of the `!` operator knows what to do when the operand has conditional state, and we just introduced a new way that expressions can have conditional state after visiting.

333fred reviewed (approved):
LGTM (commit 6)
